[자동차 경주 리팩터링] 호두 미션 제출합니다.

Junyoung-WON opened the pull request:
# 자동차 경주 미션 - Step 2
제이든 님께서 리뷰를 진행해주셨습니다.
Step 1에서의 피드백을 반영하여 미션을 구현하였습니다.

## 수정 및 변경점

### 기능 명세 세부 작성
구현할 기능 목록들을 더 세분화하여 구체적으로 작성해보았습니다.
모호하게 분류되어 있던 기능도 적절히 구분하여 기능 목록을 보기 쉽게 수정했습니다.

### 커밋 메세지 상세 표기
커밋 메세지가 중복되는 경우를 줄이고자, 커밋의 내용을 더 상세하게 작성하였습니다.
추가로 설명하고자 하는 내용이 있는 경우 제목이 아닌 본문 부분에 작성하였습니다.

### end-of-file 오류 해결
파일의 끝(코드의 마지막 커서 위치) 부분에 개행 문자가 없는 경우 나타나는 에러 메세지입니다.
보통 아래 두 가지 이유로 나타납니다.
- 운영체제가 다른 경우 개행을 표기하는 문자가 다르기 때문에,  오류가 발생할 수 있습니다.
- UNIX에서 정의한 하나의 행에 대한 정의에 맞지 않아 컴파일 시 오류가 발생할 수 있습니다.
문제 해결을 위해 IntelliJ 설정에서 파일의 끝에 자동으로 개행을 추가하도록 설정하였습니다.

### 불필요한 인스턴스화를 object를 이용하여 제거
InputView, OutputView, ValidationUtil, WinnerService 인스턴스를 GameController가 꼭 갖고 있어야 할 필요는 없습니다.
해당 클래스의 멤버 메서드만을 활용하기 때문에, 싱글톤 방식의 사용이 더 적절할 수 있습니다.
코틀린에서 싱글톤 방식을 지원하는 키워드인 object 키워드를 사용하여 InputView, OutputView, ValidationUtil, WinnerService 클래스를 수정하였습니다.
 그리하여 위 4개의 클래스에 대한 불필요한 인스턴스 생성을 없앴습니다.

### runCatching 메서드 활용
코틀린에서 지원하는 예외 처리 메서드인 runCatching 메서드를 try-catch 대신 활용해보았습니다.
GameController 내 getCarNames와 getRoundCounts 메서드에서 while문을 없애고,
runCatching내에서 자동차 이름 입력, 시도 횟수 입력에 대한 예외가 발생한 경우 getOrElse를 통해 예외 처리를 수행 후 재귀 함수 호출로 해당 입력을 반복하도록 하였습니다.

### 테스트코드 작성
사용자의 입력(자동차 이름 입력, 시도 횟수 입력)의 유효성을 검사하는 테스트 코드 - GameControllerTest
자동차의 이동 규칙에 대한 테스트 코드 - CarTest
우승자가 여럿 존재하는 경우에 대한 테스트 코드 - WinnerServiceTest
3가지 테스트 클래스를 작성하여 기능 테스트를 진행해보았습니다.

### 자동차 경주 클래스 생성하여 MVC 패턴에 맞게 리팩터링 진행
RacingCar 클래스를 생성하여 기존에 Controller가 자동차와 레이싱 경주를 관리하던 것을, RacingGame 클래스가 경주와 자동차의 움직임을 관장하고 컨트롤러는 명령만 내릴 수 있도록 변경하였습니다.

## 구현하지 못한 피드백
### 추상화(functional interface 활용)
테스트 코드를 작성하기 용이하도록, 그리고 도메인에서 굳이 난수 생성 기능을 가지지 않도록 추상화 키워드를 활용할 수 있다고 피드백 해주신 부분을 구현하지 못했습니다ㅠㅠ 조금 더 공부해보고 사용법을 올바르게 터득한 후에 활용하고 싶어서 해당 부분은 시간 문제로 반영할 수 없었습니다.

## 질문 사항
피드백을 바탕으로 코드를 리팩토링 하면서 몇가지 궁금증이 생겼습니다. 
1. runCatching을 이번 리팩터링 때에 활용하여 코드를 수정해보았습니다. 작성한 코드가 runCatching을 적절히 사용하고 있는 코드인지 알고 싶습니다.
2. runCatching은 가독성이 높고 다양한 예외 상황을 처리할 수 있다는 장점을 가지고 있는 것으로 이해하였습니다. 그렇다면 항상 runCatching을 사용하는 것이 좋을까요? 이번 미션에서 runCatching을 사용하려 했을 때, Result 타입의 활용이 조금 애매했다고 생각했습니다. 상황에 따라 반환할 값이 꼭 필요하다는 점이, 사용자의 잘못된 입력 시 다시 입력받도록 반복해야 한다는 점과 잘 맞지 않다고 느꼈습니다. runCatching을 활용하기 좋은 경우는 어떤 경우인지 궁금해졌습니다.

sorkwhwo committed:
docs: 기능 목록 작성

Junyoung-WON committed:
refactor: InputView 클래스를 객체 선언 방식으로 변경

object 키워드로 싱글톤 구현 및 사용

Junyoung-WON committed:
refactor: OutputView 클래스를 객체 선언 방식으로 변경

object 키워드를 사용하여 싱글톤 구현 및 사용

Junyoung-WON committed:
refactor: ValidationUtil 클래스를 객체 선언 방식으로 변경

object 키워드를 사용하여 싱글톤 구현 및 사용

Junyoung-WON committed:
refactor: WinnerService 클래스를 객체 선언 방식으로 변경

object 키워드를 사용하여 싱글톤 구현 및 사용

Junyoung-WON committed:
refactor: getCarNames의 runCatching 메서드 사용과 재귀 함수로의 변경

runCatching 메서드 사용, while 문 대신 재귀 사용

Junyoung-WON committed:
refactor: getRoundCounts의 runCatching 메서드 사용과 재귀 함수로의 변경

runCatching 메서드 사용, while 문 대신 재귀 사용

Junyoung-WON committed:
test: 자동차 이름 글자 수에 대한 테스트코드 추가

자동차 이름은 1~5글자 이내여야 한다.

Junyoung-WON committed:
test: 자동차 이름 형식에 대한 테스트코드 추가

자동차 이름은 한글, 영어, 숫자만 가능하며, 특수문자 등의 다른 문자는 사용할 수 없다.

Junyoung-WON committed:
test: 시도 횟수 입력 대한 테스트

시도 횟수 입력 값은 자연수이어야 한다. 숫자가 아닌 다른 문자가 오거나 0보다 작은 수는 입력할 수 없다.

Junyoung-WON committed:
refactor: 무작위 숫자를 뽑는 코드의 위치 수정

테스트 코드 작성을 위한 무작위 숫자의 생성 코드의 위치를 Controller 내 playRacingGame 메서드 내부로 수정

Junyoung-WON committed:
test: 자동차 move 메소드 테스트 코드 작성

자동차는 뽑은 숫자가 4 이상인 경우 이동한다. 4보다 작은 숫자인 경우 이동하지 않는다.

Junyoung-WON committed:
test: 우승자가 여러 명인 경우의 테스트 코드 작성

이동 거리가 가장 큰 자동차가 여럿 있는 경우 우승자는 해당 자동차 모두이다.

KwonDae commented on src/main/kotlin/racingcar/controller/GameController.kt:
```diff
@@ -23,42 +19,44 @@ class GameController {
     private fun playRacingGame(cars: List<Car>, roundCounts: Int) {
-        outputView.printResultMessage()
+        OutputView.printResultMessage()
         repeat(roundCounts) {
             cars.forEach {
-                it.move()
+                it.move(Random.nextInt(10))
             }
-            outputView.printRoundResult(cars)
+            OutputView.printRoundResult(cars)
         }
     }
```
Controller역할을 하고있는 GameCpmtroller에 대해서 한번 고민해볼까요?

<img width="1075" alt="image" src="https://github.com/woowacourse/kotlin-racingcar/assets/33443660/4e231879-4628-400e-abf4-34a2878a03df">

Controller 는 Domain(model)과 상호작용을 합니다.
그리고 domain에는 여러개의 경주차량을 관리하는 경주장(RacingGame)과 여러개의 차량(car)들로 이루어져 있죠.

그런데 지금은 경주장이 아닌 컨트롤타워에서 경주에 대한 상태, 우승자 판단 같은 것들을 혼자 다 맡아서 하고 있습니다.
때문에 테스트에서도 domain 비즈니스 로직 검증이 아닌 race 의 함수를 호출하고 있죠

즉, 명령은 Controller 담당입니다.
경주장에 대한 상태 관리와 들어온 명령에 대한 판단은 경주장 몫입니다.
차량은 자신이 얼마나 움직일지, 얼만큼 움직였는지에 대한 책임을 갖습니다.

controller는 말그대로 명령만 내리고, 레이싱게임에 관련된 정보를 관리하는건 domain(model)입니다.
RacingGame이라는 게임판을 관리하는 모델이 있다고 가정하면, 이 모델에서 우승자를 가려야 하고, 차를 움직이기도 해야합니다.
즉, 차를 움직이라는 명령을 내림
Controller -> domain(model) 흐름이 흘러가야 하는거죠.

지금 Car라는 차 객체를 어디서 관리하고 있는지 한번 고민해보시길 바랄게요!

KwonDae commented on src/main/kotlin/racingcar/service/WinnerService.kt:
```diff
@@ -2,9 +2,9 @@ package racingcar.service
-class WinnerService {
+object WinnerService {
     fun findWinners(cars: List<Car>): List<String> {
         val maxPosition = cars.maxOfOrNull { it.position }
         return cars.filter { it.position == maxPosition }.map { it.name }
```
kotlin.collections 기본 기능 활용 👍

KwonDae commented on src/main/kotlin/racingcar/controller/GameController.kt:
```diff
@@ -23,42 +19,44 @@ class GameController {
         repeat(roundCounts) {
             cars.forEach {
-                it.move()
+                it.move(Random.nextInt(10))
```
Controller는 명령만 내릴 뿐 비즈니스 로직에는 관여하지 않습니다.
지금은 Random.nexInt(10)이라는 랜덤값과, 범위도 Controller가 정해주고 있네요.

아래 사진 구조를 보며 같이 고민해볼까요?

체스판을 예로 들어 볼게요,
체스를 플레이하는 유저는 Controller가 될 수 있을것이고,
체스판은 RacingGame이 될 수 있겠죠. 체스판에는 현재 체스판의 상태(누가 이기고 있는지, 어떤 말들이 죽고 살았는지 등등) 상태를 갖고있습니다.
유저는 단지 체스판에게 어떤 말을 얼마나 움직이라고 명령할 뿐이죠

현실세계에 대입하여 한번 고민해보시죠!

KwonDae replied:
https://techcourse.woowahan.com/s/Q0BXzOWp/ls/hWTE4dpQ
위 구조에서 보면 랜덤값을 가져오는 로직이 Domain(model)에 종속되어야 할까요?

지금 랜덤으로 숫자를 가져오는 부분이 kotlin.random.Random 에 의존하고 있어 해당 getRandomNumber 함수와 이를 사용하는 곳에서는 테스트가 어렵습니다.

만약 정해진 숫자가 나온다면 함수 호출 후의 상태 변화 혹은 결과가 예측되기에 테스트를 쉽게 작성할 수 있는데요,
어떻게 하면 Car 모델에서 getRandomNumber 메소드가 Random 에 의존하지 않고 테스트 가능하게끔 할 수 있을까요?

이럴떄 추상화라는 개념을 사용하곤 합니다. 실제로 domain 에서는 추상화를 통한 로직을 호출하고, 이를 통해 의존성을 분리하여 테스트를 가능하게끔 할 수 있죠. (interface를 테스트하는곳에서 override를 할 수 있다!)

우리는 여기서 [Functional Interface](https://kotlinlang.org/docs/fun-interfaces.html)를 활용해볼 수 있는데요
바로 아래 피드백과 연관성이 크네요. 함꼐 고민해보시고 어려운점 있으시다면 언제든 편하게 질문주세요!

추가로, random 요소들을 테스트 하기 쉬운 구조로 개선하는 방법에 대해
아래 패턴에 대해서 참고해보시면 좋을것 같습니다 👍
[전략패턴1](https://victorydntmd.tistory.com/292)
[전략패턴2](https://refactoring.guru/ko/design-patterns/strategy)

KwonDae reviewed (changes requested):
안녕하세요 호두!

2단계 미션도 깔끔하게 잘 해주셨습니다.
MVC패턴에 대해 코멘트를 몇개 드렸으니 확인해보시고
꼭 다시 요청 부탁드립니다!
끝까지 화이팅입니다 💪

KwonDae commented:
> runCatching을 이번 리팩터링 때에 활용하여 코드를 수정해보았습니다. 작성한 코드가 runCatching을 적절히 사용하고 있는 코드인지 알고 싶습니다.

깔끔하게 runCatching - getOrElse 를 잘 활용해주셨습니다 👍 
특히 runCatching은 어떠한 결과가 success/fail 로 결과값을 받아와야할 때 빛을 발하는데요, onSuccess/onFailure 로 분기문에 대한 처리도 가능하고, 작성해주신것 처럼 getOrElse 로 예외처리도 가능합니다.

> runCatching은 가독성이 높고 다양한 예외 상황을 처리할 수 있다는 장점을 가지고 있는 것으로 이해하였습니다. 그렇다면 항상 runCatching을 사용하는 것이 좋을까요? 이번 미션에서 runCatching을 사용하려 했을 때, Result 타입의 활용이 조금 애매했다고 생각했습니다. 상황에 따라 반환할 값이 꼭 필요하다는 점이, 사용자의 잘못된 입력 시 다시 입력받도록 반복해야 한다는 점과 잘 맞지 않다고 느꼈습니다. runCatching을 활용하기 좋은 경우는 어떤 경우인지 궁금해졌습니다.

느끼신점이 정확합니다. 무조건적으로 runCatching 이 좋다? ❌❌
각 상황에 맞게 사용하는게 정답입니다.
[try-catch와 비교해놓은 글](https://www.baeldung.com/kotlin/runcatching-vs-try-finally)이 있으니 참고해보시고, 각자 쓰임새가 어떻게 다른지 확인해보세요!

Junyoung-WON committed:
refactor: 자동차 객체가 스스로 난수를 생성하고 그에 따라 이동하도록 변경

자동차 클래스가 스스로의 이동을 책임지도록 수정.

Junyoung-WON committed:
refactor: 자동차들을 관리하고 라운드를 진행하는 RacingGame 클래스를 도메인에 추가

MVC 패턴에 맞게 리팩터링: 기존에 GameController에서 자동차들을 관리하던 것을 RacingGame이라는 클래스 내에서 자동차를 관리(경주 수행)하도록 변경

Junyoung-WON committed:
refactor: 불필요한 패키지 importing 제거

KwonDae commented on src/main/kotlin/racingcar/domain/RacingGame.kt:
```diff
@@ -0,0 +1,11 @@
+package racingcar.domain
+
+class RacingGame(carNames: List<String>) {
+    val cars: List<Car> = carNames.map { Car(it) }
+
+    fun playOneRound() {
+        cars.forEach {
+            it.move()
+        }
+    }
+}
```
Domain(model)을 잘 만들어 주셨어요!
이제 여기서 controller로부터 명령을 받아 우승자를 계산하는 로직과 같이 어떤게 들어오면 좋을지 고민해봐도 좋을것 같습니다 👍

KwonDae commented on src/main/kotlin/racingcar/domain/Car.kt:
```diff
@@ -6,7 +6,8 @@ class Car(val name: String) {
-    fun move(number: Int) {
+    fun move() {
+        val number: Int = Random.nextInt(10)
         if (number >= 4) position++
```
이제 자동차 스스로 움직일지 말지 결정하게 되었네요!
그렇다면 이 `move()` 메소드를 단위테스트 가능하도록 하기 위해 어떻게 하면 좋을지 고민해보면 좋겠습니다.

KwonDae reviewed (approved):
2단계 미션까지 정말 고생많으셨습니다 👍 
1단계보다 책임 분리가 되면서 보다 확장성있고 단위테스트가 더 많이 가능하도록 되었습니다.

코멘트 드린 부분들 잘 반영해주시느라 정말 고생많으셨어요 🙂
남겨드린 코멘트는 참고해보시고 추가 문의사항 있으시면 언제든 편하게 DM 주시면 됩니다.

본 미션은 여기서 마무리하겠습니다.
앞으로도 화이팅입니다 💪
